Modules/_winapi.c

```diff
@@ -774,12 +774,134 @@ gethandle(PyObject* obj, const char* name)
     return ret;
 }
 
+static PyObject* sortenvironmentkey(PyObject *module, PyObject *item) {
+    return _winapi_LCMapStringEx_impl(NULL, LOCALE_NAME_INVARIANT, LCMAP_UPPERCASE, item);
+}
+
+static PyMethodDef sortenvironmentkey_def = {
+    "sortenvironmentkey",
+    _PyCFunction_CAST(sortenvironmentkey),
+    METH_O,
+    ""
+};
+
+static PyObject *
+normalize_environment(PyObject* environment)
+{
+    PyObject *result = NULL, *keys = NULL, *keyfunc = NULL, *sort = NULL,
+             *args = NULL, *kwargs = NULL;
+
+    keys = PyMapping_Keys(environment);
+    if (keys == NULL) {
+        return NULL;
+    }
+
+    keyfunc = PyCFunction_New(&sortenvironmentkey_def, NULL);
+    if (keyfunc == NULL) {
+        goto error;
+    }
+    sort = PyObject_GetAttrString(keys, "sort");
+    if (sort == NULL) {
+        goto error;
+    }
+    args = PyTuple_New(0);
+    if (args == NULL) {
+        goto error;
+    }
+    kwargs = PyDict_New();
+    if (kwargs == NULL) {
+        goto error;
+    }
+    if (PyDict_SetItemString(kwargs, "key", keyfunc) < 0) {
+        goto error;
+    }
+    if (PyObject_Call(sort, args, kwargs) == NULL) {
+        goto error;
+    }
+
+    result = PyDict_New();
+    if (result == NULL) {
+        goto error;
+    }
+    wchar_t *prev_key_string = NULL;
+
+    for (int i=0; i<PyList_GET_SIZE(keys); i++) {
+        PyObject *key = PyList_GET_ITEM(keys, i);
+        if (key == NULL) {
+            goto error;
```

Comment:
You are missing a `Py_DECREF(result)` here (and some other places).

Instead of adding many statements like that, you could also just add a `Py_XDECREF(result)` to the `error` section. And add a`result=NULL` at the end of the `error` section.

Reply:
Updated, according to what you described here, there needs to be a way to determine whether we are in the error path or success path. I have used `if (PyErr_Occurred())` as the flag, and it seems to work fine. However, I haven't come across this usage in the CPython codebase. Is this the correct usage?

Reply:
I did not think this through completely. `PyErr_Occurred` is used extensively within cpython, but I am not sure this works here (not all failing methods will set the error flag).
Looking at the code right now the best way seems to be to add a `Py_DECREF(result); result=0; goto error;` to each failure clause. It is a bit repetitive (which I tried to avoid), but then the result can just be returned at the end of the method.

Reply:
Thanks for the reply, updated.